titus-api/src/test/java/com/netflix/titus/api/jobmanager/model/job/validator/ParallelValidatorTest.java

```diff
@@ -0,0 +1,232 @@
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+/**
+ * This class tests the {@link ParallelValidator} class.
+ */
+public class ParallelValidatorTest {
+    private static final JobDescriptor MOCK_JOB = mock(JobDescriptor.class);
+    private static final Duration TIMEOUT = Duration.ofMillis(10);
+
+    // Hard validation tests
+
+    @Test
+    public void validateHardPassPass() {
+        EntityValidator pass0 = new PassJobValidator();
+        EntityValidator pass1 = new PassJobValidator();
+        ParallelValidator validator = new ParallelValidator(TIMEOUT, Arrays.asList(pass0, pass1), Collections.emptySet());
+        Mono<Set<ValidationError>> mono = validator.validate(MOCK_JOB);
+
+        StepVerifier.create(mono)
+                .expectNext(Collections.emptySet())
+                .verifyComplete();
+    }
+
+    @Test
+    public void validateHardFailFail() {
+        EntityValidator fail0 = new FailJobValidator();
+        EntityValidator fail1 = new FailJobValidator();
+        ParallelValidator validator = new ParallelValidator(TIMEOUT, Arrays.asList(fail0, fail1), Collections.emptySet());
+        Mono<Set<ValidationError>> mono = validator.validate(MOCK_JOB);
+
+        StepVerifier.create(mono)
+                .expectNextMatches(errors -> errors.size() == 2)
+                .verifyComplete();
+
+        Set<ValidationError> errors = mono.block();
+        validateFailErrors(errors);
+        validateErrorType(errors, ValidationError.Type.HARD);
+    }
+
+    @Test
+    public void validateHardPassFail() {
+        EntityValidator pass = new PassJobValidator();
+        EntityValidator fail = new FailJobValidator();
+        ParallelValidator validator = new ParallelValidator(TIMEOUT, Arrays.asList(pass, fail), Collections.emptySet());
+        Mono<Set<ValidationError>> mono = validator.validate(MOCK_JOB);
+
+        StepVerifier.create(mono)
+                .expectNextMatches(errors -> errors.size() == 1)
+                .verifyComplete();
+
+        Set<ValidationError> errors = mono.block();
+        validateFailErrors(errors);
+        validateErrorType(errors, ValidationError.Type.HARD);
+    }
+
+    @Test
+    public void validateHardPassTimeout() {
+        EntityValidator pass = new PassJobValidator();
+        EntityValidator never = new NeverJobValidator();
+        EntityValidator validator = new ParallelValidator(TIMEOUT, Arrays.asList(pass, never), Collections.emptySet());
+        Mono<Set<ValidationError>> mono = validator.validate(MOCK_JOB);
+
+        StepVerifier.create(mono)
+                .expectNextMatches(errors -> errors.size() == 1)
+                .verifyComplete();
+
+        Set<ValidationError> errors = mono.block();
```

Comment:
Why this check is not above in the expectNextMatches?

Reply:
Why not below?  Ordering is unimportant here.